File: studio/services/publish_service.py
"""Mock publish pipeline."""

import time
import uuid

from studio import db
from studio.models import Quality, Release, to_doc
from studio.quality.engine import run_quality
from studio.repositories import analytics_repo, profiles_repo, working_html_repo


def run_quality_gate(project_id):
    profile = profiles_repo.get(project_id)
    html = working_html_repo.require_html(project_id)
    result = run_quality(html, profile)
    with db.session() as s:
        s.add(
            Quality(projectId=project_id, result=result, ts=time.time())
        )
        s.commit()
    return result


def latest_quality(project_id):
    with db.session() as s:
        row = (
            s.query(Quality)
            .filter_by(projectId=project_id)
            .order_by(Quality.ts.desc())
            .first()
        )
        return row.result if row else None


def publish(project_id, *, channel="preview", version_id=None):
    gate = run_quality_gate(project_id)
    if gate["verdict"] == "fail":
        raise ValueError("quality gate failed")
    release_id = uuid.uuid4().hex[:12]
    row = Release(
        releaseId=release_id,
        projectId=project_id,
        channel=channel,
        versionId=version_id,
        status="success",
        createdAt=time.time(),
    )
    with db.session() as s:
        s.add(row)
        s.commit()
        doc = to_doc(row)
    analytics_repo.track(project_id, "publish_success", metadata={"channel": channel})
    return doc


def list_releases(project_id):
    with db.session() as s:
        rows = (
            s.query(Release)
            .filter_by(projectId=project_id)
            .order_by(Release.createdAt.desc())
            .all()
        )
        return [to_doc(r) for r in rows]

File: src/server/services/publishService.ts
import * as analyticsRepo from "../repos/analyticsRepo";
import * as profilesRepo from "../repos/profilesRepo";
import * as workingHtmlRepo from "../repos/workingHtmlRepo";
import { Prisma } from "@/generated/prisma/client";
import { runQuality } from "./contextService";
import { prisma } from "../db";
import { toDoc } from "../repos/doc";

export async function runQualityGate(projectId: string): Promise<Record<string, unknown>> {
  const profile = await profilesRepo.get(projectId);
  const html = await workingHtmlRepo.requireHtml(projectId);
  const result = runQuality(html, profile);
  await prisma().builderQuality.create({
    data: { projectId, result: result as Prisma.InputJsonValue, ts: Date.now() / 1000 },
  });
  return result;
}

export async function latestQuality(projectId: string): Promise<Record<string, unknown> | null> {
  const row = await prisma().builderQuality.findFirst({
    where: { projectId },
    orderBy: { ts: "desc" },
  });
  return (row?.result as Record<string, unknown> | null) ?? null;
}

export async function publish(
  projectId: string,
  channel = "preview",
  versionId: string | null = null,
): Promise<Record<string, unknown>> {
  const gate = await runQualityGate(projectId);
  if (gate.verdict === "fail") throw new Error("quality gate failed");
  const releaseId = randomId();
  const row = await prisma().builderRelease.create({
    data: {
      releaseId,
      projectId,
      channel,
      versionId,
      status: "success",
      createdAt: Date.now() / 1000,
    },
  });
  await analyticsRepo.track(projectId, "publish_success", { channel });
  return toDoc(row) as Record<string, unknown>;
}

export async function listReleases(projectId: string): Promise<Record<string, unknown>[]> {
  const rows = await prisma().builderRelease.findMany({
    where: { projectId },
    orderBy: { createdAt: "desc" },
  });
  return rows.map((row) => toDoc(row) as Record<string, unknown>);
}

function randomId(): string {
  const buf = new Uint8Array(6);
  crypto.getRandomValues(buf);
  return Array.from(buf, (b) => b.toString(16).padStart(2, "0")).join("");
}
